proof/crefine/lib/Boolean_C.thy

```diff
@@ -0,0 +1,138 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2014, General Dynamics C4 Systems
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+(* Definitions and rules for C boolean constants parsed from the kernel *)
+
+theory Boolean_C
+imports
+  "CSpec.KernelInc_C"
+begin
+
+lemma true_and_1[simp]:
+  "true && 1 = true"
+  "signed true && 1 = signed true"
+  "unsigned true && 1 = unsigned true"
+  by (simp add: true_def)+
+
+lemma to_bool_true[simp]:
+  "to_bool true"
+  "to_bool (signed true)"
+  "to_bool (unsigned true)"
+  by (simp add: to_bool_def true_def)+
+
+lemma true_equals_simps[simp]:
+  "true = 1 \<longleftrightarrow> True"
+  "signed true = 1 \<longleftrightarrow> True"
+  "unsigned true = 1 \<longleftrightarrow> True"
+  "true = 0 \<longleftrightarrow> False"
+  "signed true = 0 \<longleftrightarrow> False"
+  "unsigned true = 0 \<longleftrightarrow> False"
+  by (simp add: true_def)+
+
+lemma true_eq_from_bool[simp]:
+  "(true = from_bool P) = P"
+  "(signed true = from_bool P) = P"
+  "(unsigned true = from_bool P) = P"
+  by (simp add: from_bool_def split: bool.splits)+
+
+lemma false_and_1[simp]:
+  "false && 1 = false"
+  "signed false && 1 = signed false"
+  "unsigned false && 1 = unsigned false"
+  by (simp add: false_def)+
+
+lemma to_bool_false[simp]:
+  "\<not> to_bool false"
+  "\<not> to_bool (signed false)"
+  "\<not> to_bool (unsigned false)"
+  by (simp add: to_bool_def false_def)+
+
+lemma false_equals_simps[simp]:
+  "false = 0 \<longleftrightarrow> True"
+  "signed false = 0 \<longleftrightarrow> True"
+  "unsigned false = 0 \<longleftrightarrow> True"
+  "false = 1 \<longleftrightarrow> False"
+  "signed false = 1 \<longleftrightarrow> False"
+  "unsigned false = 1 \<longleftrightarrow> False"
+  by (simp add: false_def)+
+
+lemma false_eq_from_bool[simp]:
+  "(false = from_bool P) = (\<not> P)"
+  "(signed false = from_bool P) = (\<not> P)"
+  "(unsigned false = from_bool P) = (\<not> P)"
+  by (simp add: from_bool_def split: bool.splits)+
+
+lemma from_bool_vals[simp]:
+  "from_bool True = signed true"
+  "from_bool False = signed false"
```

Comment:
Do we also have `unsigned` versions of these? They might already be in the simp set.

Reply:
These were moved from `IpcCancel_C` now that we have a central location for them to go in. I don't think `unsigned` versions of them exist.

Reply:
Would the unsigned version just be `from_bool True = true`? In any case, it's fine to merge as is, this is something we could add in another PR if we wanted to.

Reply:
Hmm, I've now managed to confuse myself with this. I don't think that we want `from_bool True = true`, because that would only apply for `from_bool :: bool ⇒ 32 signed word`.

In some sense I don't think that it really matters, in that we're basically picking a form to simplify the `from_bool` into and the `signed`/`unsigned` is mostly to get the types right.

Reply:
Thinking about it a bit more, it's possible that quite a few of the lemmas in this file could have the repetition around `unsigned` and `signed` reduced. Because `true` has a  specific value and type, we can show that `signed true = unsigned true` holds and pick whichever form we prefer.

Reply:
Agreed, I think this is Ok.